test/boost/types_test.cc

```diff
@@ -107,6 +107,13 @@ BOOST_AUTO_TEST_CASE(test_ascii_type_string_conversions) {
     test_parsing_fails(ascii_type, "¡Hola!");
 }
 
+BOOST_AUTO_TEST_CASE(test_utf8_type_string_conversions) {
+    BOOST_REQUIRE(utf8_type->equal(utf8_type->from_string("utf8"), utf8_type->decompose("utf8")));
+    BOOST_REQUIRE_EQUAL(utf8_type->to_string(utf8_type->decompose("utf8")), "utf8");
+
+    // test_parsing_fails(ascii_type, ""); <- something to fail validation
+}
```

Comment:
I don't understand what this test is testing... And what is that commented part?
I would expect a good test to "prove" your patch: The test should fail before your patch and demonstrate the bug, and pass after your patch.

Reply:
Moreover, the title of the pull request suggests that this patch fixes validation for *CQL*. But this test doesn't check CQL, and it can't convince me CQL isn't still broken. Remember that we *thought* it wasn't broken also before, because the checking code existed!
To check it works in CQL you can write a test in `cql_query_test.cc` or if you prefer we also have the two newer CQL testing mechanisms - `test/cql` and `test/cql-pytest`.

Reply:
The testing part is not yet there as this was just a quick draft I made to mark it as in progress in the discussion and will complete that today (as mentioned there as well)

Reply:
Good. I didn't realize this was a "draft" pull request (now I see where it says it... It's not very conspicuous).
May I ask what's the point of a "draft pull request" of this sort? Who is supposed to use it or review it?

Reply:
As for the CQL part, I mirrored my previous commit fixing that for ascii types. The name should probably be a bit more suited to what it actually fixes, which is broader (validation for from string conversions)
If there is concern, I can also add a CQL test explicitly as well if we want.

Reply:
I think if the change is done and only tests are missing then it makes sense to show the code to the reviewers while working on tests.

Reply:
Draft PR == RFC Patch

Reply:
> As for the CQL part, I mirrored my previous commit fixing that for ascii types. The name should probably be a bit more suited to what it actually fixes, which is broader (validation for from string conversions)
> If there is concern, I can also add a CQL test explicitly as well if we want.

Well, if I understood correctly, this bug was first discovered in CQL, and after some digging you managed to hunt it down to this particular change to this internal data structure. So it's good to have a unit test that verifies this internal data structure, but I think it's also good (and perhaps even more important) to verify that you fixed the real user-visible CQL bug.

Reply:
Maybe we could test the change with CQL test?

Reply:
> Maybe we could test the change with CQL test?

I hereby volunteer to write one, in test/cql-pytest (in Python). I hope I'll have a ready patch for your review very soon.

Reply:
This is really weird - the test I wrote works without this patch - when I try to insert broken utf-8 to a 'text' type I get:
```
E   cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid HEX constant (0xce80) for "t" of type text"
```
So when does this bug exist?

Reply:
Hm, I can confirm that it's working for me as well
@avikivity you reported that it's broken, how can we reproduce that?

Reply:
What's the test?

I did not reproduce myself, but saw from the code that `INSERT INTO tab(foo) VALUES('broken utf8')` will succeed, but should fail.

Reply:
@avikivity I just sent to the mailing list a test which checks exactly that - inserting broken utf8 into a "text" - and it already works correctly, i.e., fails as expected, with exactly the same error message as Cassandra (something like "Invalid HEX constant (...) for "t" of type text".). So maybe you looked at the wrong code.

In any case, let's not accept this patch before understand what exactly it fixes. In what code path is this function that you're changing actually used? if you comment out this code entirely, what doesn't compile? :-)

Reply:
For me this fails: `INSERT INTO ks.invalid (id, t) VALUES (2, 0xC0);`, where t is of type `text`
I wonder, if what you saw was indeed failing and this one is succeeding, maybe our validation is invoked properly, but doesn't cover some specific corner cases or smth

Are you able to find what was the broken utf8 exactly?

Reply:
`saw from the code ...` xD

Reply:
> For me this fails: `INSERT INTO ks.invalid (id, t) VALUES (2, 0xC0);`, where t is of type `text`
> I wonder, if what you saw was indeed failing and this one is succeeding, maybe our validation is invoked properly, but doesn't cover some specific corner cases or smth
> 
> Are you able to find what was the broken utf8 exactly?

@StarostaGit how did you run this command? Is this the command verbatim - with 0xC0 in the string and not the character 0xC0?

Reply:
As I'm thinking about it now, maybe the important question is whether a text value put in ' ' should be a raw string or something that accounts for special chars (after `\` for example, like `\x60`)

Reply:
@StarostaGit when you say "for me this fails", I guess you mean the insert fails, i.e., Scylla succeeds, behaves as expected - right?

Reply:
> @StarostaGit how did you run this command? Is this the command verbatim - with 0xC0 in the string and not the character 0xC0?

I checked with "0xC0" (I didn't know CQL had this syntax, nice, I can update my test to also check that) and the insert fails as expected. I guess I misunderstood you when you said "for me this fails" - the insert fails, so Scylla works correctly. It seems there is no bug here.

Reply:
> For me this fails: `INSERT INTO ks.invalid (id, t) VALUES (2, 0xC0);`, where t is of type `text`
> I wonder, if what you saw was indeed failing and this one is succeeding, maybe our validation is invoked properly, but doesn't cover some specific corner cases or smth
> 

This is invoking a different path. `0xC0` isn't a string literal, it's a blob literal, so it's probably rejected earlier.

Reply:
Yes, I meant that it "fails as expected"
Sorry for the confusion

Reply:
@avikivity can you suggest a way to pass an invalid string literal rather than a blob literal?
Also, why would a blob literal be "rejected earlier"? 0xC0 is perfectly fine as a blog... Only when it is handled as a string should it be rejected.

Reply:
0xC0 is a fine blob, but a blob is not a string. It should be rejected in the same way as `SET text_column = 3` or `SET numeric_column = 'nadav'` is rejected.

Reply:
Probably via an `assignment_testable::is_assignable` failure.

Reply:
```c++
bytes
constants::literal::parsed_value(data_type validator) const
{
    try {
        if (_type == type::HEX && validator == bytes_type) {
            auto v = static_cast<sstring_view>(_text);
            v.remove_prefix(2);
            return validator->from_string(v);
        }
        if (validator->is_counter()) {
            return long_type->from_string(_text);
        }
        return validator->from_string(_text);
    } catch (const marshal_exception& e) {
        throw exceptions::invalid_request_exception(e.what());
    }
}
```

Reply:
Actually see `constants::literal::test_assignment`.

Reply:
I did also blobastext(0xC0), but this unfortunately fails to convert too soon to have any meaning here (as expected tho)
However, at least the path that was wrongly handled in the ascii case seems to be correct here - I added a test that is trying to build a `utf8_type` from a given string (in `boost/types_test.cc`):

```
void test_parsing_fails(const shared_ptr<const abstract_type>& type, sstring str)
{
    try {
        type->from_string(str);
        BOOST_FAIL(format("Parsing of '{}' should have failed", str));
    } catch (const marshal_exception& e) {
        // expected
    }
}

// other stuff, tests, all that jazz and the 

BOOST_AUTO_TEST_CASE(test_utf8_type_string_conversions) {
    test_parsing_fails(utf8_type, "\xC0\xC1");
}
```

Reply:
> 0xC0 is a fine blob, but a blob is not a string. It should be rejected in the same way as `SET text_column = 3` or `SET numeric_column = 'nadav'` is rejected.

It is in fact not rejected, neither in Cassandra nor Scylla. But maybe indeed it goes through a different path. The question is how can you NOT go through that different path in a test. I'm still trying to come up with a test in CQL, not a unit tests involving our internal types. I'll let you know if I find something.

Reply:
Without the code snippet that you mentioned should've been failing but wasn't I'm afraid we might be wandering in the dark here

Reply:
I'm sorry, I had a few bugs in my original test which made it *seem* like I found a bunch of bugs in CQL's UTF-8 validation, when in fact there are none. The new version of my tests are in this PR: https://github.com/scylladb/scylla/pull/7719
All the tests now pass on Scylla, meaning there is no visible (for these tests) bug in CQL's UTF-8 validation. Which of course we doesn't mean we don't have a bug in some internal code path that isn't reached by any of the CQL commands I tested (I actually tested 4 different CQL requests that need to part UTF-8 strings).
So now that I have a much better understanding of this issue, I'll re-review your patches.